### ISSUE TITLE
setTimeout Issues in WebView2

### ISSUE THREAD

**Issue by RickStrahl** [MicrosoftEdge/WebView2Feedback#1595]
I'm back at trying to get the bloody Alt-key Menu forwarding to work in WPF after I thought I had that working, it keeps giving me problems.

Scenario:

The problem I'm solving with the code below is to handle Alt key forwarding to the WPF host app so the menu Alt key processing works - ie. alt key activates the menu and then further keystrokes navigate the menu. The problem is that the WebView - unlike the WebBrowser Control - doesn't handle this key forwarding automatically. I've set up code in JavaScript to intercept the Alt key, check for a delay between keys and then change focus from the WV to the WPF host and simulate the alt key press. It sort of works - but there's a problem with the event handling in the control it appears.

The following capture demonstrates the funky behavior.

Basically I handle the `keyDown` event, then set a marker to let me know the alt key was pressed, then `setTimeout()` to delay a 300 milliseonds to check if another key was pressed in between. If not I forward the key to .NET and the WPF host. The logic is fine it works - but not consistently.

The problem is that `setTimeout()` does not fire many times - **until I perform some other UI action** - another keypress or a mouse click.

![setTimeoutWeirdness](https://user-images.githubusercontent.com/1374013/127782387-95fa1023-adeb-46e2-a642-1ec266a91405.gif)

In the capture I start by focusing in my editor, then pressing `alt`. You can see the key handler trigger, and the alt key being active. However, note that the second console.log (in the `setTimeout()` does not trigger) - until I move the mouse and click or type another key.

I repeat the same scenario a second time.

Then after I've actually activated the menu at least once it sometimes works fine. The last couple of times you can see pressing alt and all the events fire and the menu activates.

I double checked my code in the rest of the function and commented out any other code that fires into .NET/WPF so there's nothing blocking on the Interop end from my application - the key is passed through like any other key otherwise. 

Yet for some reason the `setTimeout()` call is blocked until some sort of UI interaction occurs which seems very wrong and likely attributable to the hosting scenario. 

This feels like a typical 'DoEvents()` type operation that's stuck waiting for some other event that never happens.

Any ideas what would cause this and has this been seen before?

**Comment by champnic** [comment]
Hey @RickStrahl - I haven't seen this before, and I'm not sure what would cause this. It's surprising. I think if the app is in the background or hidden some events get throttled so I wonder if that's what's happening here. If you have this code snippet in the browser (Edge or Chrome) do you see a similar behavior?

For what it's worth I agree that our key handling is definitely lacking here and making these scenarios way more complicated than they should be. Fixing keyboard handling is one of our top upcoming priorities and I hope we can make improvements here soon. Thanks!

**Comment by RickStrahl** [comment]
Yeah if I run this just in the browser it works fine - the setTimeout always triggers. I just can't think of another reason why this wouldn't fire than some odd interaction between the WebView and the host UI since it frees up as soon as a mouse move or other UI interaction occurs. Oddly other events are firing - it's not like there's a block, with another key being accepted. If I type another key which hits this same event handler again it fires immediately. 

Something is preventing the setTimeout() from firing until some other UI activation occurs.

**Comment by champnic** [comment]
Hey @RickStrahl - Would you be able to take a trace and send that to us?

```
wpr -start EdgeBrowser
<run the scenario>
wpr -stop EdgeBrowser.etl
```

You can email the resulting EdgeBrowser.etl it directly to me at `champnic@microsoft.com` (or upload to cloud and send me a link), and then we should be able to take a closer look.

Thanks!

**Comment by RickStrahl** [comment]
Tried to do this but I can't get it to work:

![image](https://user-images.githubusercontent.com/1374013/130525859-c20f94f8-be43-4ed2-aa04-a3c7157a3a8a.png)

Always fails when trying to save. Running either as Admin or normal User.

I'm not sure if this would help in this scenario. I've got this set to tight scenario - editor open ready to press Alt key once (with no effect) - second alt-key then triggers (because of the UI activation) and it works. Not sure of what you would see there that would let you see why the setTimeout is not firing unless a UI event triggers it...

**Comment by champnic** [comment]
Sounds like the `wpr -stop` bug might be fixed in newer versions:
"Please use the insider version of ADK Windows Performance Toolkit, build 19650 or later"
from: https://devblogs.microsoft.com/performance-diagnostics/wpr-start-and-stop-commands/

**Comment by mungojam** [comment]
If it is due to background 'tab' throttling, I really hope there is a switch to turn that off. I was hoping to move our LOB app from PWA to webview in order to avoid the throttling which currently impacts our users

**Comment by harunurhan** [comment]
@champnic we have the same issue using WinForms, when Alt is pressed I think focus is moved out (to system menu I assume?) and this causes event queues to be stuck, so none of the timeout and I believe also promise callbacks to hold forever to run unless you focus on something. Surprisingly even focusing on another window fixes the issue.

Indeed this could be due to background tab throttling that chromium also has, as called out by @mungojam.

Has anyone tried `disable-background-timer-throttling`? or would that be even an acceptable workaround?

**Comment by harunurhan** [comment]
Update: I am able to reproduce the issue even with `--disable-background-timer-throttling --disable-renderer-backgrounding --disable-renderer-backgrounding` flags passed as additional browser arguments :(

Actually using below on the `Form` which hosts the WebView2 control seems to be working.

```c#
MenuStrip hiddenMenu = new MenuStrip()
{
    Visible = false,
};
this.MainMenuStrip = hiddenMenu;
this.Controls.Add(hiddenMenu);
```

However with this workaround, you lose the ability to access the system menu using the `alt` then `arrow-down` keys, but access still works using `alt + spacebar` combo. This may be acceptable for some apps, but some apps do need the menu strip and even when you don't need it, there might be other consequence of this hack.

<img width="180" height="157" alt="Image" src="https://github.com/user-attachments/assets/2e011a83-d8a7-4a5c-bdc3-e963f9d70488" />

**Comment by harunurhan** [comment]
Btw I can reproduce this bug consistently on Teams too, so priority might need to be bumped up @champnic
